### PR TITLE
feat(icon-toggle): Add initial implementation.

### PR DESCRIPTION
**Pull request by traviskaufman**
Additionally, refactored some common test helpers into their own
`helpers/` dir.

Part of #4470
[Delivers #126819521]

### CODE REVIEW

File: packages/mdl-icon-toggle/README.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
+# MDL Icon Toggle
+
+MDL Icon Toggle provides a material design icon toggle button. It is fully accessible, and is
```
Comment (traviskaufman) [inline packages/mdl-icon-toggle/README.md]: That's a great question. It shouldn't according to the [W3C spec](https://www.w3.org/TR/wai-aria/roles#textalternativecomputation):

Algorithm rule `2A`:

> If aria-labelledby is empty or undefined, the aria-label attribute, which defines an explicit text string, is used...

Algorithm rule `2C`:

> Otherwise, if the attributes checked in rules A and B didn't provide results, text is collected from descendant content if the current element's role allows "Name From: contents."

Comment (Garbee) [inline packages/mdl-icon-toggle/README.md]: We also need to remember to note ligatures are not required to be used. Codepoints can also be used just fine (and required to avoid some issues like Google Translate breaking a page) in which case we should apply the `aria-label` attribute to define the help text.

File: packages/mdl-icon-toggle/index.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    super.destroy();
+  }
+
+  getDefaultFoundation() {
+    return new MDLIconToggleFoundation({
+      addClass: className => this.iconEl_.classList.add(className),
+      removeClass: className => this.iconEl_.classList.remove(className),
+      registerInteractionHandler: (type, handler) => this.root_.addEventListener(type, handler),
+      deregisterInteractionHandler: (type, handler) => this.root_.removeEventListener(type, handler),
+      setText: text => {
```
Comment (sgomes) [inline packages/mdl-icon-toggle/index.js]: Nit: drop `{}`.

File: packages/mdl-icon-toggle/index.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        this.iconEl_.textContent = text;
+      },
+      getTabIndex: () => /* number */ this.root_.tabIndex,
+      setTabIndex: tabIndex => {
```
Comment (sgomes) [inline packages/mdl-icon-toggle/index.js]: Nit: drop `{}`.

File: packages/mdl-icon-toggle/mdl-icon-toggle.scss [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  @include mdl-theme-dark(".mdl-icon-toggle", true) {
+    @include mdl-theme-prop(color, text-disabled-on-dark);
+  }
+}
+
+/** postcss-bem-linter: end */
+
+.mdl-icon-toggle.mdl-ripple-upgraded {
+  @include mdl-ripple-base;
+  // NOTE: The spec denotes specific opacity values to use for different styles of icon buttons,
```
Comment (sgomes) [inline packages/mdl-icon-toggle/mdl-icon-toggle.scss]: Comment of the day.

File: packages/mdl-icon-toggle/mdl-icon-toggle.scss [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  @include mdl-ripple-bg((pseudo: "::before", opacity: .062));
+  @include mdl-ripple-fg((pseudo: "::after", opacity: .062));
+
+  @include mdl-theme-dark(".mdl-icon-toggle", true) {
+    @include mdl-ripple-bg((pseudo: "::before", base-color: white, opacity: .16));
+    @include mdl-ripple-fg((pseudo: "::after", base-color: white, opacity: .16));
+  }
+
+  // For some reason, GPU layer promotion makes ripples look terrible on Chrome
+  will-change: initial;
```
Comment (sgomes) [inline packages/mdl-icon-toggle/mdl-icon-toggle.scss]: Is performance OK on mobile without its own layer?

File: test/unit/mdl-checkbox/foundation.test.js [hunk 6]
```diff
@@ -18,6 +18,9 @@ import test from 'tape';
 import bel from 'bel';
 import lolex from 'lolex';
 import td from 'testdouble';
+
+import {setupFoundationTest} from '../helpers/setup';
```
Comment (sgomes) [inline test/unit/mdl-checkbox/foundation.test.js]: More checkbox changes creeping in.

**Comment by sgomes** [comment]
Also, the build seems to be failing with what looks like a broken test, not flakiness.

File: packages/mdl-checkbox/mdl-checkbox.scss [hunk 7]
```diff
@@ -18,9 +18,6 @@
 @import "./variables";
 @import "./keyframes";
 
-// TODO: Docs
-// TODO: x-browser test!
-
 @function mdl-checkbox-transition($property, $timing-function, $delay: 0ms, $duration: $mdl-checkbox-transition-duration) {
```
Comment (traviskaufman) [inline packages/mdl-checkbox/mdl-checkbox.scss]: Ah yeah I figured since I was looking at that file anyway I'd take the time to remove some outdated TODOs. I can update the commit msg and/or remove those changes if you want. Let me know!

File: packages/mdl-icon-toggle/README.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+designed to work with any icon set.
+
+## Installation
+
+> NOTE: Installation via npm will be available post-alpha.
+
+## Usage
+
+In order to use MDL Icon Toggle, you will need to import an icon font, such as [Material Icons](https://design.google.com/icons/) or [Font Awesome](http://fontawesome.io/).
```
Comment (traviskaufman) [inline packages/mdl-icon-toggle/README.md]: Good point! I think I meant "icon set".

File: packages/mdl-icon-toggle/README.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+`mdl-icon-toggle--accent` that allow you to color mdl-icon-toggle based on your primary and accent
+colors, respectively.
+
+### Listening for change events
+
+`MDLIconToggle` emits an `MDLIconToggle:change` custom event when the value of the icon toggle
+changes _as a result of user input_. This decision was made to align with how `change` events work
+for normal inputs. In addition, these events do not bubble and cannot be cancelled.
+
+The custom event's `detail` object contains a property `isOn` denoting whether or not the ripple is
```
Comment (traviskaufman) [inline packages/mdl-icon-toggle/README.md]: Yup totally right. That 100% reflects the status of the icon toggle. Not sure why I wrote that...great catch!

File: test/unit/mdl-checkbox/foundation.test.js (hunk 6 above)
Comment (traviskaufman) [inline test/unit/mdl-checkbox/foundation.test.js]: Yeah similar to above I figured I'd tackle some refactors along with this PR so that we can minimize tech debt while pushing out features at the same time. 

Should I split these cleanup changes into a separate commit and then we could merge the PR in without squashing the commits?

File: packages/mdl-icon-toggle/README.md [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+```js
+const iconEl = document.querySelector('.mdl-icon-toggle');
+const status = document.getElementById('status');
+iconEl.addEventListener('MDLIconToggle:change', ({detail}) => {
+  status.textContent = `Icon Toggle is ${detail.isOn ? 'on' : 'off'}`;
+});
+```
+
+### Refreshing the toggle data via the vanilla component.
```
Comment (traviskaufman) [inline packages/mdl-icon-toggle/README.md]: I think I'm just going to get rid of this altogether and just parse whenever the method is called. It avoids a micro-optimization and makes the code a lot simpler.

File: packages/mdl-icon-toggle/index.js (hunk 3 above)
Comment (traviskaufman) [inline packages/mdl-icon-toggle/index.js]: eslint complains about this, and rightfully so:

screen shot 2016-09-21 at 14 59 36

This will actually return a value as the result of an expression. Although it doesn't make a difference, it seems more straightforward to explicitly wrap this statement in a BlockExpression. If you disagree, we can disable the rule?

File: packages/mdl-icon-toggle/index.js (hunk 2 above)
Comment (traviskaufman) [inline packages/mdl-icon-toggle/index.js]: see comment below

File: packages/mdl-icon-toggle/mdl-icon-toggle.scss (hunk 5 above)
Comment (traviskaufman) [inline packages/mdl-icon-toggle/mdl-icon-toggle.scss]: just checked and it seems okay. The pseudo-els are still promoted into their own layer so that should be sufficient for compositing.

**Comment by traviskaufman** [comment]
@sgomes changes made PTAL

File: packages/mdl-checkbox/mdl-checkbox.scss (hunk 7 above)
Comment (sgomes) [inline packages/mdl-checkbox/mdl-checkbox.scss]: I'm OK with small changes like this, as long as we don't think it's a problem for our commit strategy, particularly where it comes to automatic changelogs.

File: packages/mdl-icon-toggle/README.md (hunk 8 above)
Comment (sgomes) [inline packages/mdl-icon-toggle/README.md]: "icon set" SGTM

File: packages/mdl-icon-toggle/index.js (hunk 2 above)
Comment (sgomes) [inline packages/mdl-icon-toggle/index.js]: Good point. Let's leave as-is, then.

File: packages/mdl-icon-toggle/index.js (hunk 3 above)
Comment (sgomes) [inline packages/mdl-icon-toggle/index.js]: Good point. Let's leave as-is, then.

File: packages/mdl-icon-toggle/mdl-icon-toggle.scss (hunk 5 above)
Comment (sgomes) [inline packages/mdl-icon-toggle/mdl-icon-toggle.scss]: Sweet, thanks!

File: test/unit/mdl-checkbox/foundation.test.js (hunk 6 above)
Comment (sgomes) [inline test/unit/mdl-checkbox/foundation.test.js]: Again, I'm OK with either splitting or not, depending on whether you think it's a problem for our commit message strategy.

**Comment by sgomes** [comment]
PS: You may want to take a look at TravisCI before merging, though. It looks like the build is failing on IE11.
